### PR TITLE
libnetwork/driverapi: make discoverAPI an optional part of the interface

### PR DESCRIPTION
**Pull request by thaJeztah**
### libnetwork/driverapi: make discoverAPI an optional part of the interface

Most drivers do not implement this, so detect if a driver implements
the discoverAPI, and remove the implementation from drivers that do
not support it.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
@corhere @akerouanton happy to hear your thoughts on this one. I can squash (or drop) commits, depending on which of the implementations makes most sense (kept them separate, as I went a bit back-and-forth on what made most sense)

File: libnetwork/drvregistry/networks.go [hunk 1]
```diff
@@ -9,7 +9,7 @@ import (
 )
 
 // DriverWalkFunc defines the network driver table walker function signature.
-type DriverWalkFunc func(name string, driver driverapi.Driver, capability driverapi.Capability) bool
+type DriverWalkFunc func(name string, driver driverapi.Driver, capability driverapi.Capability) (stop bool)
```
Comment (corhere) [inline libnetwork/drvregistry/networks.go:12]: This change is out of scope for this PR

**Review by corhere — COMMENTED** [review]
I would have done this myself if you hadn't gotten to it. I only ask that we have a compile-time assertion that overlay and the other remaining driver implementations do implement `discoverapi.Discover`.
```go
var _ discoverapi.Discover = (*driver)(nil)
```

File: libnetwork/drvregistry/networks.go (hunk 1 above)
Comment (thaJeztah) [reply libnetwork/drvregistry/networks.go:12]: Ah, yeah, I had to look up what the return was 😂

File: libnetwork/drivers/remote/driver_test.go [hunk 2]
```diff
@@ -472,14 +472,16 @@ func TestRemoteDriver(t *testing.T) {
...
 
-	data := discoverapi.NodeDiscoveryData{
-		Address: "192.168.1.1",
-	}
-	if err = d.DiscoverNew(discoverapi.NodeDiscovery, data); err != nil {
-		t.Fatal(err)
-	}
-	if err = d.DiscoverDelete(discoverapi.NodeDiscovery, data); err != nil {
-		t.Fatal(err)
+	if dr, ok := d.(discoverapi.Discover); ok {
```
Comment (thaJeztah) [inline libnetwork/drivers/remote/driver_test.go:475]: This change can be dropped once https://github.com/moby/moby/pull/46101 is merged

**Review by corhere — APPROVED** [review]

File: libnetwork/drivers/remote/driver_test.go (hunk 2 above)
Comment (thaJeztah) [reply libnetwork/drivers/remote/driver_test.go:475]: done 👍
